src/ripple/ledger/Directory.h

```diff
@@ -0,0 +1,163 @@
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_LEDGER_DIR_H_INCLUDED
+#define RIPPLE_LEDGER_DIR_H_INCLUDED
+
+#include <ripple/ledger/ReadView.h>
+#include <ripple/protocol/Indexes.h>
+
+namespace ripple {
+
+class Dir
+{
+private:
+    ReadView const* view_ = nullptr;
+    Keylet root_;
+
+    // memoization
+    std::shared_ptr<SLE const> sle_ = nullptr;
+    STVector256 const* indexes_ = nullptr;
+
+public:
+    class iterator;
+    using value_type = std::shared_ptr<SLE const>;
+
+    Dir(ReadView const&, Keylet const&);
+
+    Dir() = delete;
+
+    Dir(Dir const&) = default;
+
+#if defined(_MSC_VER) && _MSC_VER <= 1800
+    Dir(Dir&& other);
+
+    Dir&
+    operator=(Dir&& other);
+
+#else
+    Dir(Dir&&) = default;
+
+    Dir&
+    operator=(Dir&&) = default;
+
+#endif
+
+    Dir&
+    operator= (Dir const&) = default;
+
+    iterator
+    begin() const;
+
+    iterator
+    end() const;
+
+    iterator
+    find(uint256 const& page_key, uint256 const& sle_key) const;
+};
+
+class Dir::iterator
+{
+public:
+    using value_type =
+        Dir::value_type;
+    using pointer =
+        value_type const*;
+    using reference =
+        value_type const&;
+    using difference_type =
+        std::ptrdiff_t;
+    using iterator_category =
+        std::forward_iterator_tag;
+
+    iterator() = default;
+
+    iterator(iterator const&) = default;
+
+    iterator(ReadView const& view,
+        Keylet const& root,
+        Keylet const& page)
+    : view_(&view)
+    , root_(root)
+    , page_(page)
+    {
+    }
+
+#if defined(_MSC_VER) && _MSC_VER <= 1800
+    iterator(iterator&& other);
+
+    iterator&
+    operator= (iterator&& other);
+
+#else
+    iterator(iterator&&) = default;
+
+    iterator&
+    operator= (iterator&&) = default;
+
+#endif
+
+    iterator&
+    operator=(iterator const& other);
+
+    bool
+    operator==(iterator const& other) const;
+
+    bool
+    operator!=(iterator const& other) const
+    {
+        return ! (*this == other);
+    }
+
+    reference
+    operator*() const;
+
+    pointer
+    operator->() const { return &**this; }
+
+    iterator&
+    operator++();
+
+    iterator
+    operator++(int);
+
+    Keylet const&
+    page() const { return page_; }
+
+    uint256
+    index() const { return index_; }
+
+private:
+    friend class Dir;
+
+    ReadView const* view_ = nullptr;
+    Keylet root_;
+    Keylet page_;
+    uint256 index_;
+    boost::optional<value_type> mutable cache_;
+
+    // memoization
+    std::shared_ptr<SLE const> sle_ = nullptr;
```

Comment:
In general, it means to remember. sle_ and indexes_ are assigned from relatively expensive calls that we can avoid recalculating often. These members serve only as an optimization.

cache_ primarily is required in dereference as a local would leave scope.